### DISCUSSION TITLE
JavaScript sdk and core - uploading files is weird in pocketbase.

### DISCUSSION THREAD

**Discussion by briosheje** [pocketbase/pocketbase#2004] [Q&A] [answered]
I've been extensively using PocketBase in the last months and had overall a very pleasant experience with it, but I wanted to raise a question about what design decision brought the team to the decision of using FormData to upload a file to PocketBase.

To clarify, I usually work with services that to upload a file takes these steps:

1) Upload the file somewhere.
2) Get the URL of the file back.
3) Use the URL of the file to reference the file itself.

What I see in PocketBase instead is:

1) Pick a file.
2) Upload the entire payload using FormData and the file itself.

It seems cleverer and shorter, but these issues are going to come up:

- If you're used to work with JSON payloads, you need to rewrite all your logic for your DTO to be sent to PocketBase, because you need to transfer the ENTIRE data object with FormData instead.
- If you have any sort of abstraction of the UI, it is going to be painful to handle such.
- As a consequence of the first point of this list, you need to write your own logic to convert your existing JSON payload to FormData.

Of course, that's a design decision, but I was wondering whether there is anything I'm missing with such approach that I didn't find very clever and pretty much against any file uploading practice I've seen before.

I do know that doing the upload in such a way prevents handling unused files, dangling files and all of that stuff, but that seems a bit odd to me to have to write all the logic to handle FormData when there just shouldn't be the need to do so.

In case that's helpful to anyone, here is a simple function that converts a generic model to a FormData object that can be directly used with the PocketBase SDK. I'm not sure it handles all the case, but I've been using it extensively and didn't have any edge case uncovered with it yet:

```ts
export default function modelToFormData<TModel extends {}>(model: TModel) {
  const outFormData = new FormData();

  for (const [k, v] of Object.entries(model)) {
    if (Array.isArray(v)) {
      for (const value of v) {
        outFormData.append(k, value);
      }
    } else if (v instanceof Date) {
      outFormData.append(k, v.toISOString());
    } else if (v === undefined) {
      // Do nothing if the value is undefined.
    } else if (
      v instanceof File ||
      v === null ||
      typeof v === "number" ||
      typeof v === "string" ||
      typeof v === "boolean" ||
      typeof v === "bigint"
    ) {
      outFormData.append(k, v as string);
    } else if (typeof v === "object") {
      outFormData.append(k, JSON.stringify(v));
    } else {
      console.error(model);
      throw new Error(
        `Unknown value type provided. value: ${v}, type: ${typeof v}`
      );
    }
  }

  return outFormData;
}
```

**Comment by ganigeorgiev** [comment, marked as answer]
I don't understand what is the issue with `FormData`.

Your described usual workflow is effectively doing the same. 

> 1. Upload the file somewhere.
> 2. Get the URL of the file back.
> 3. Use the URL of the file to reference the file itself.

About 1) - How you are uploading the file? Via FormData, right?

About 2) and 3) - What happens if the record has an invalid regular field (eg. missing title)? Do you manually delete the already uploaded file or there is now need for an additional service that will cleanup the unlinked files after some custom timeout?

Being able to send the file(s) in the same request as your regular fields guarantees that the uploaded files are stored only if the record data are also persisted.

Of course, if you you want to continue following your existing workflow, you can:
- send 2 separate requests - one with only the uploaded files and second request only with the regular fields
- AND/OR create a separate "media/files" collection where you'll upload the files first (this is similar to your first point) and then link the returned record id in your main record with a `relation` field (this is similar to 2 and 3).

**Reply by ganigeorgiev** [reply]
> If you're used to work with JSON payloads, you need to rewrite all your logic for your DTO to be sent to PocketBase, because you need to transfer the ENTIRE data object with FormData instead.

Eventually to make it slightly easier, we can allow passing `File/Blob` values as regular json body fields and do the `FormData` conversion behind the scene before sending the request, but I'll have to think a little more on it since it could have some unintended side-effects.

**Reply by briosheje** [reply]
Please don't get anything wrong from my considerations: I'm not saying what is right or wrong, I'm asking whether some design decisions were taken, and why.

As far as I could see, PocketBase does some linking between the uploaded file and the record related to it, so I do understand why it is in this way.

The sdk also mentions that out of box it automatically handles the request format (handles multipart/formdata in case a FormData object is passed), but what I usually see diffused the most nowaday is sending plain JSON objects to servers with datatype application/json and see the files uploaded somewhere else.

I'm not saying the pocketbase approach is wrong, I'm asking why it works in a such a way and whether there is some other approach I couldn't foresee when I did my implementation.

I do like the approach of handling files manually (the one you proposed), but still, this goes in contrast with another thing that I personally had problems with: URLs of files are calculated through the PB library client side. This is another thing that I've never seen before and in my case I needed to make an adapter for it, but that's fine.

To clarify (once again): I'm not asking for any library change, I'm just curious about why such implementation was chosen as in my experience I've never seen any SaaS using such an approach.

Updating the library to handle json to a FormData automatically would be a great feature in my opinion, although I do understand that this could have some side effects and that requires some time and test, but I would really think about including such, since you already are supporting both FormData and Object inputs in the sdk itself.

**Reply by ganigeorgiev** [reply]
> what I usually see diffused the most nowaday is sending plain JSON objects to servers with datatype application/json and see the files uploaded somewhere else.

But again, you still need to send a separate `FormData` request with the file. PocketBase allows you to do both depending on your use case or your preference.

Other services handles the storage separately because most of the time it is implemented as an independent microservice (eg. for better reliability and easier scaling).

> URLs of files are calculated through the PB library client side. This is another thing that I've never seen before and in my case I needed to make an adapter for it, but that's fine.

There is no magic in the file url generation. The available JS SDK helper (aka. `pb.getFileUrl`) just generate a string in the format `/api/files/:collectionIdOrName/:recordId/:filename`.
We store only the filename as field value because it is a easier to migrate/change the internal storage implementation without having to update the database.

**Reply by briosheje** [reply]
I know no magic is happening about the URL generation, but once again I usually get back the full URL file from the request and don't have to apply any transformation to it, but that's feasible and not an original argumentation. The question I had was what design decision brought you to implement such feature in a single request whereas many other providers do separate requests and I think the answer is that since it's technically a monolith and there is no service that manages the file system it makes in fact sense to do everything in a single request, that's what I was looking for.

Although it makes sense as it currently is, I would really think that adding some automation about that in the SDK would make such implementation much simpler, but once again I do understand that it means potentially having side effects.

**Comment by denisgurev** [comment]
I like how pocketbase handles files. Please don't change it :pray:.

It is a lot easier compared to firebase and as @ganigeorgiev wrote, you can always create a dedicated files collection if you prefer the firebase way.

**Reply by briosheje** [reply]
I'm not asking to change it, I'm asking why it was done in this way.
